Readme.md
# k

  keyboard event dispatcher.

## Installation

    $ component install yields/k

## Features

  - `super` keyword is `command` on mac and `ctrl` on anything else.
  - gmail like sequences `(a b c *)`
  - key combos
  - nice simple api
  - you can unbind anything you bind

## Example

```js
var k = require('k')(window);

// combos
k('command + shift + enter', fn);

// on OSX `super` will be replaced with `command`
// on other operating systems it will be replaced with `ctrl`.
k('super + k', fn);

// sequences
k('a b c', fn);
k('* b c', fn);
k('super + shift + a b c', fn);
```

## API

### k(el)

  Create a new keyboard dispatcher on the given `element`

### k(keys, fn)

  Bind the given `fn` on `keys`.

### .el

  the element `k` is listening on.

### #bind

  Bind `keys` with `fn(e)`.

### #unbind

  Unbind all listeners of `keys`,
  unbind `fn` with `keys`,
  unbind all listeners on all keys.

```js
k.unbind(); // => unbind all listeners
k.unbind('enter'); // => unbind all enter listeners
k.unbind('command + enter'); // => unbind all `command + enter` listeners
k.unbind('command + enter', fn); // => unbind `fn` just from `command + enter`
```

### #destroy

  Removes all listeners on `k.el`.
  Removes all `k.listeners`.

### #ignore

  this function is called just before any handlers are called,
  if it returns `true` no handlers will be called for this event.

  by default it returns `true` for `input`, `textarea` or `select`.

  if you still want `k` to handle `inputs` you can `k.ignore = false`.

### .shift

  `true` if `shiftKey` is down.

### .ctrl

  `true` if `ctrlKey` is down.

### .alt

  `true` if `alt` is down.

### .command

  `true` if `command` is down.

### .modifiers

  `null` if no modifiers are down, otherwise it's `true`.

### .super

  `true` if `command` or `ctrl` are down (osx / others)

## Tests

```bash
$ make test
```

## License

  MIT


History.md

0.7.2 / 2015-03-19
==================

  * fix tests
  * better reset on focus

0.7.0 / 2014-12-01
==================

  * modifiers: consider inactive modifiers
  * sequences: support cmd + *
  * package: add "repository" field
  * package: rename to "yields-k" for npm
  * component: add "browser" keyword

0.6.2 / 2014-05-19
==================

 * fix for older browsers and reserved keywords (<IE9)

0.6.1 / 2014-03-03
==================

 * fix for firefox keycode, closes #21

0.6.0 / 2013-10-30
==================

 * add gmail like sequences, closes #17
 * add keyup support, closes #11
 * use array for listeners

0.5.2 / 2013-10-27
==================

 * fix bug when unbinding multiple listeners [karlbohlmark]

0.5.1 / 2013-10-10
==================

 * fix unbind("mod + key")

0.5.0 / 2013-09-28
==================

 * refactor
 * Better `.unbind()` [dandean]

0.4.1 / 2013-09-28
==================

 * fix .unbind(keys, fn), closes #13

0.4.0 / 2013-09-16
==================

 * add package.json
 * remove "merge" dep
 * add "super" support, closes #6
 * add multiple modifiers support

0.3.0 / 2013-08-16
==================

 * clean unbind
 * typo: break, instead of continue

0.2.1 / 2013-06-08
==================

 * add ie7-9 support
 * left command keycode is 93 [lepture]

0.2.0 / 2013-04-28
==================

  * add comma support
  * add .modifiers, closes #3

0.1.3 / 2013-04-28
==================

  * fix modifier clearing, closes #4

0.1.2 / 2013-04-27
==================

  * remove global keycode() from tests

0.1.1 / 2013-04-27
==================

  * keycode dep
  * keycode dep fix [lepture]

0.1.0 / 2013-04-25
==================

  * ignore keypress on input, select and textarea
  * move proto to ./proto.js

0.0.1 / 2013-01-31
==================

  * readme typo
  * initial commit


test/k.js

var dispatcher = require('k');
var os = require('os');

describe('k', function(){

  // superkey

  var superkey = 'mac' == os
    ? 'command'
    : 'ctrl';

  // keycode

  function keycode(name){
    switch (name) {
      case ',': return 188;
      case 'command': return 91;
      case 'shift': return 16;
      case 'ctrl': return 17;
      case 'alt': return 18;
      case 'enter': return 13;
      default: return name[0].toUpperCase().charCodeAt(0);
    }
  }

  // keyname

  function keyname(code){
    return String.fromCharCode(code).toLowerCase();
  }

  // press `el` with `code`

  function press(el, code){
    if (!el || !code) throw new Error('(el, code) must be given to press()');
    code = keycode(code);
    var e = document.createEvent('Event');
    e.initEvent('keydown', true, true);
    e.keyCode = e.which = code;
    el.dispatchEvent(e);
    return function(){
      var e = document.createEvent('Event');
      e.initEvent('keyup', true, true);
      e.keyCode = code;
      el.dispatchEvent(e);
    };
  }

  // create element with `type`

  function elem(type){
    return document.createElement(type || 'div');
  }

  // assert `expr` or throw `ms`

  function assert(expr, ms){
    if (expr) return;
    throw new Error(ms || 'oh noes!');
  }


  describe('k = k(el)', function(){
    it('should create a new dispatcher with `el`', function(){
      var el = elem();
      var k = dispatcher(el);
      assert(k.el === el);
      assert(dispatcher(el) !== k);
    })
  })

  describe('k.super', function(){
    it('should be "' + superkey + '" on "' + os + '"', function(){
      assert(superkey == dispatcher(window).super);
    })
  })

  describe('k(keys, fn)', function(){
    it('should bind the given `fn` on `keys`', function(){
      var k = dispatcher(elem());
      k('shift + enter', function(){});
      k('a, b, c', function(){});
      assert('shift' == k.listeners[13][0].mods[0]);
      assert(k.listeners[keycode('a')]);
      assert(k.listeners[keycode('b')]);
      assert(k.listeners[keycode('c')]);
    })

    it('should support bind `,` key currectly', function(){
      var k = dispatcher(elem());
      k(',', function(){});
      assert(k.listeners[keycode(',')]);
    })

    it('should support keys like `command + ,, ctrl + ,`', function(){
      var k = dispatcher(elem());
      k('ctrl + ,, command + ,', function(){});
      assert(2 == k.listeners[keycode(',')].length)
      assert('ctrl' == k.listeners[keycode(',')][0].mods[0]);
      assert('command' == k.listeners[keycode(',')][1].mods[0]);
    })

    it('should support multiple modifers `command + shift + ,`', function(){
      var k = dispatcher(elem());
      k('command + shift + ,', function(){});
      assert(2 == k.listeners[keycode(',')][0].mods.length);
    })
  })

  describe('k(e)', function(){
    it('should invoke all listeners that match the event', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('enter', function(){ ++invoked; });
      k('a', function(){ ++invoked; });
      k('b', function(){ ++invoked; });
      k('c', function(){ ++invoked; });
      press(el, 'enter')();
      press(el, 'a')();
      press(el, 'b')();
      press(el, 'c')();
      assert(4 == invoked);
    })

    it('should invoke keys like `ctrl + ,, command + ,` and `,`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      function incr(){ ++invoked; }
      k(',', incr);
      k('ctrl + ,, command + ,', incr);
      press(el, ',')();
      assert(1 == invoked);
      var unpress = press(el, 'ctrl');
      press(el, ',')();
      unpress();
      assert(3 == invoked);
      unpress = press(el, 'command');
      press(el, ',')();
      unpress();
      assert(5 == invoked);
    })

    it('should work with multiple modifers `command + shift + ,`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      function incr(){ ++invoked; }
      k('command + shift + ,', incr);
      press(el, 'command');
      assert(0 == invoked);
      press(el, 'shift');
      assert(0 == invoked);
      press(el, ',');
      assert(1 == invoked);
    })

    it('should work with modifiers like `shift + enter, ctrl + enter`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      var unpress, mods = ['shift', 'ctrl', 'alt', 'command'];
      k('shift + enter', function(){ ++invoked; });
      k('ctrl + enter', function(){ ++invoked; });
      k('alt + enter', function(){ ++invoked; });
      k('command + enter', function(){ ++invoked; });
      for (var i = 0; i < mods.length; ++i) {
        unpress = press(el, mods[i]);
        press(el, 'enter')();
        assert(1 + i == invoked);
        unpress();
      }
    })

    it('should set `.MODIFIER` to `true` when its down', function(){
      var el = elem();
      var k = dispatcher(el);
      var mods = ['ctrl', 'shift', 'command', 'alt'];
      var unpress;
      for (var i = 0; i < mods.length; ++i) {
        unpress = press(el, mods[i]);
        assert(true == k[mods[i]]);
        unpress();
        assert(null == k[mods[i]])
      }
    })

    it('should not clear all modifiers onkeyup unless they are not down', function(){
      var el = elem();
      var k = dispatcher(el);
      var mods = ['ctrl', 'shift', 'command', 'alt'];
      var unpress;
      for (var i = 0; i < mods.length; ++i) {
        unpress = press(el, mods[i]);
        press(el, 'enter')();
        assert(true == k[mods[i]]);
        unpress();
        assert(null == k[mods[i]]);
      }
    })

    it('should ignore input, select and textarea elements by default', function(){
      var el = elem('input');
      var k = dispatcher(el);
      var invoked = 0;
      k('enter', function(){ ++invoked; });
      k('a', function(){ ++invoked; });
      press(el, 'enter')();
      press(el, 'a')();
      assert(0 == invoked);
    })
  })

  describe('k.modifiers', function(){
    it('should be `true` if a modifier(s) is down', function(){
      var el = elem();
      var k = dispatcher(el);
      var mods = ['ctrl', 'shift', 'command', 'alt'];
      var unpress;
      for (var i = 0; i < mods.length; ++i) {
        unpress = press(el, mods[i]);
        assert(true == k.modifiers);
        unpress();
        assert(null == k.modifiers);
      }
    })

    it('should be `falsey` if no modifiers are down', function(){
      var el = elem();
      var k = dispatcher(el);
      assert(null == k.modifiers);
      var shift = press(el, 'shift');
      var ctrl = press(el, 'ctrl');
      assert(true == k.modifiers);
      shift();
      assert(true == k.modifiers);
      ctrl();
      assert(null == k.modifiers);
    })
  })

  describe('k("super + enter")', function(){
    it('should be invoked on "' + superkey + ' + enter"', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('super + enter', function(){ ++invoked; });
      var sup = press(el, superkey);
      var enter = press(el, 'enter');
      assert(1 == invoked);
      sup();
      enter();
    })

    it('should set "' + superkey + '" and .super to true when super is down', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('super + enter', function(){ ++invoked; });
      var sup = press(el, superkey);
      assert(k[superkey]);
      assert(k.super);
      sup();
    })
  })

  describe('k.unbind([keys, [fn]])', function(){
    it('should unbind the given `fn` from `keys`', function(){
      var k = dispatcher(elem());
      k('shift + enter', console.log);
      k('shift + enter', console.dir);
      assert(2 == k.listeners[13].length);
      k.unbind('enter', console.dir);
      assert(1 == k.listeners[13].length);

      k('shift + 1', console.warn);
      assert(1 == k.listeners[49].length);
      k.unbind('shift + 1', console.warn);
      assert(0 == k.listeners[49].length);
    })

    it('should unbind all listeners if `fn` is omitted', function(){
      var k = dispatcher(elem());
      k('shift + enter', assert);
      k('shift + enter', assert);
      assert(2 == k.listeners[13].length);
      k.unbind('enter');
      assert(0 == k.listeners[13].length);
    })

    it('should reset listeners if no arguments are given', function(){
      var k = dispatcher(elem());
      k('enter', assert);
      k('a', assert);
      k('b', assert);
      assert(k.listeners[13]);
      assert(k.listeners[keycode('a')]);
      assert(k.listeners[keycode('b')]);
      k.unbind();
      assert(!k.listeners[keycode('b')]);
      assert(!k.listeners[keycode('a')]);
      assert(!k.listeners[13]);
    })
  })
})


test/test.js

describe('k', function(){

  var keycode = require('keycode')
    , assert = require('assert')
    , dispatcher = require('k')
    , os = require('os');

  // superkey

  var superkey = 'mac' == os
    ? 'command'
    : 'ctrl';

  describe('(element)', function(){
    it('should create a new dispatcher with `el`', function(){
      var el = elem();
      var k = dispatcher(el);
      assert(k.el === el);
      assert(dispatcher(el) !== k);
    })
  })

  describe('(keys, fn)', function(){
    it('should bind the given `fn` on `keys`', function(){
      var k = dispatcher(elem());
      k('shift + enter', function(){});
      k('a, b, c', function(){});
      assert('shift' == k.listeners[0].mods[0]);
      assert(keycode('a') == k.listeners[1].code);
      assert(keycode('b') == k.listeners[2].code);
      assert(keycode('c') == k.listeners[3].code);
    })

    it('should support bind `,` key currectly', function(){
      var k = dispatcher(elem());
      k(',', function(){});
      assert(keycode(',') == k.listeners[0].code);
    })

    it('should support keys like `command + ,, ctrl + ,`', function(){
      var k = dispatcher(elem());
      k('ctrl + ,, command + ,', function(){});
      assert(2 == k.listeners.length)
      assert('ctrl' == k.listeners[0].mods[0]);
      assert('command' == k.listeners[1].mods[0]);
    })

    it('should support multiple modifers `command + shift + ,`', function(){
      var k = dispatcher(elem());
      k('command + shift + ,', function(){});
      assert(2 == k.listeners[0].mods.length);
    })

    it('should support sequences `a b c d`', function(){
      var k = dispatcher(elem());
      k('a b c d', function(){});
      assert('a b c d' == k.listeners[0].key);
      assert('function' == typeof k.listeners[0].seq);
      assert('function' == typeof k.listeners[0].fn);
    })

    it('should support mods with sequences `command + shift + a b c`', function(){
      var k = dispatcher(elem());
      k('command + shift + a b c', function(){});
      assert('a b c' == k.listeners[0].key);
      assert('command' == k.listeners[0].mods[0]);
      assert('shift' == k.listeners[0].mods[1]);
    })
  })

  describe('up(keys, fn)', function(){
    it('should bind a "keyup" event', function(){
      var k = dispatcher(elem());
      k.up('k', function(){});
      assert(1 == k.listeners.length);
      assert('keyup' == k.listeners[0].event);
    })

    it('should invoke only on "keyup"', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k.up('enter', function(){ ++invoked; });
      var enter = press(el, 'enter');
      assert(0 == invoked);
      enter();
      assert(1 == invoked);
    })
  })

  describe('("super + enter", fn)', function(){
    it('should be invoked on "' + superkey + ' + enter"', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('super + enter', function(){ ++invoked; });
      var sup = press(el, superkey);
      var enter = press(el, 'enter');
      assert(1 == invoked);
      sup();
      enter();
    })

    it('should work with firefox keycode for command "224"', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('super + a', function(){ ++invoked; });
      var sup = press(el, 224);
      var a = press(el, 'a');
      assert(1 == invoked);
      sup();
      a();
    })

    it('should set "' + superkey + '" and .super to true when super is down', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('super + enter', function(){ ++invoked; });
      var sup = press(el, superkey);
      assert(k[superkey]);
      assert(k.super);
      sup();
    })
  })

 describe('(event)', function(){
    it('should invoke all listeners that match the event', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('enter', function(){ ++invoked; });
      k('a', function(){ ++invoked; });
      k('b', function(){ ++invoked; });
      k('c', function(){ ++invoked; });
      k.up('d', function(){ ++invoked; });
      press(el, 'enter')();
      press(el, 'a')();
      press(el, 'b')();
      press(el, 'c')();
      var d = press(el, 'd');
      assert(4 == invoked);
      d();
      assert(5 == invoked);
    })

    it('should invoke keys like `ctrl + ,, command + ,` and `,`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      function incr(){ ++invoked; }
      k(',', incr);
      k('ctrl + ,, command + ,', incr);
      press(el, ',')();
      assert(1 == invoked);
      var unpress = press(el, 'ctrl');
      press(el, ',')();
      unpress();
      assert(2 == invoked);
      unpress = press(el, 'command');
      press(el, ',')();
      unpress();
      assert(3 == invoked);
    })

    it('should work with multiple modifers `command + shift + ,`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      function incr(){ ++invoked; }
      k('command + shift + ,', incr);
      press(el, 'command');
      assert(0 == invoked);
      press(el, 'shift');
      assert(0 == invoked);
      press(el, ',');
      assert(1 == invoked);
    })

    it('should work with modifiers like `shift + enter, ctrl + enter`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      var unpress, mods = ['shift', 'ctrl', 'alt', 'command'];
      k('shift + enter', function(){ ++invoked; });
      k('ctrl + enter', function(){ ++invoked; });
      k('alt + enter', function(){ ++invoked; });
      k('command + enter', function(){ ++invoked; });
      for (var i = 0; i < mods.length; ++i) {
        unpress = press(el, mods[i]);
        press(el, 'enter')();
        assert(1 + i == invoked);
        unpress();
      }
    })

    it('should set `.MODIFIER` to `true` when its down', function(){
      var el = elem();
      var k = dispatcher(el);
      var mods = ['ctrl', 'shift', 'command', 'alt'];
      var unpress;
      for (var i = 0; i < mods.length; ++i) {
        unpress = press(el, mods[i]);
        assert(true == k[mods[i]]);
        unpress();
        assert(null == k[mods[i]])
      }
    })

    it('should not clear all modifiers onkeyup unless they are not down', function(){
      var el = elem();
      var k = dispatcher(el);
      var mods = ['ctrl', 'shift', 'command', 'alt'];
      var unpress;
      for (var i = 0; i < mods.length; ++i) {
        unpress = press(el, mods[i]);
        press(el, 'enter')();
        assert(true == k[mods[i]]);
        unpress();
        assert(null == k[mods[i]]);
      }
    })

    it('should not invoke `super + a` if `super + shift + a` is down', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      function incr(){ ++invoked; }
      k('super + a', incr);
      var shift = press(el, 'shift');
      var cmd = press(el, 'command');
      assert(0 == invoked);
      press(el, 'a')();
      assert(0 == invoked);
      shift();
      press(el, 'a')();
      assert(1 == invoked);
    });

    it('should ignore input, select and textarea elements by default', function(){
      var el = elem('input');
      var k = dispatcher(el);
      var invoked = 0;
      k('enter', function(){ ++invoked; });
      k('a', function(){ ++invoked; });
      press(el, 'enter')();
      press(el, 'a')();
      assert(0 == invoked);
    })
  })

  describe('sequences', function(){
    it('`a b c`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('a b c', function(){ ++invoked; });
      press(el, 'a')();
      assert(0 == invoked);
      press(el, 'b')();
      assert(0 == invoked);
      press(el, 'c')();
      assert(1 == invoked);
    })

    it('`a b c` and `c`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('a b c', function(){ ++invoked; });
      k('c', function(){ ++invoked; });
      press(el, 'a')();
      assert(0 == invoked);
      press(el, 'b')();
      assert(0 == invoked);
      press(el, 'c')();
      assert(2 == invoked);
      press(el, 'c')();
      assert(3 == invoked);
    })

    it('`command + a b c`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('command + a b c', function(){ ++invoked; });
      var cmd = press(el, 'command');
      assert(0 == invoked);
      press(el, 'a')();
      press(el, 'b')();
      assert(0 == invoked);
      press(el, 'c')();
      assert(1 == invoked);
      cmd();
    })

    it('`command + *`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('command + *', function(){ ++invoked });
      var cmd = press(el, 'command');
      assert(0 == invoked);
      press(el, 'a')();
      press(el, 'b')();
      assert(2 == invoked);
      cmd();
    })

    it('`* a`', function(){
      var el = elem();
      var k = dispatcher(el);
      var invoked = 0;
      k('* a', function(){ ++invoked; });
      press(el, 'b')();
      press(el, 'a')();
      assert(1 == invoked);
      press(el, 'd')();
      press(el, 'c')();
      assert(1 == invoked);
    })
  })

  describe('#super', function(){
    it('should be "' + superkey + '" on "' + os + '"', function(){
      assert(superkey == dispatcher(window).super);
    })
  })

  describe('#unbind', function(){
    it('("shift + enter", fn)', function(){
      var k = dispatcher(elem());
      k('shift + enter', console.log);
      k('shift + enter', console.dir);
      assert(2 == k.listeners.length);
      k.unbind('enter', console.dir);
      assert(1 == k.listeners.length);

      k('shift + 1', console.warn);
      assert(2 == k.listeners.length);
      k.unbind('shift + 1', console.warn);
      assert(1 == k.listeners.length);
    })

    it('("command + enter")', function(){
      var k = dispatcher(elem());
      k('command + enter', assert);
      k('shift + enter', assert);
      assert(2 == k.listeners.length);
      k.unbind('command + enter');
      assert(1 == k.listeners.length);
    })

    it('("enter")', function(){
      var k = dispatcher(elem());
      k('shift + enter', assert);
      k('shift + enter', assert);
      assert(2 == k.listeners.length);
      k.unbind('enter');
      assert(0 == k.listeners.length);
    })

    it('("left, right")', function(){
      var k = dispatcher(elem());
      k('left', assert);
      k('right', assert);
      assert(2 == k.listeners.length);
      k.unbind('left, right');
      assert(0 == k.listeners.length);
    })

    it('("a b c")', function(){
      var k = dispatcher(elem());
      k('a b c', assert);
      assert(1 == k.listeners.length);
      k.unbind('a b c');
      assert(0 == k.listeners.length);
    })

    it('()', function(){
      var k = dispatcher(elem());
      k('enter', assert);
      k('a', assert);
      k('b', assert);
      assert(3 == k.listeners.length);
      k.unbind();
      assert(0 == k.listeners.length);
    })
  })

  describe('#modifiers', function(){
    it('should be `true` if a modifier(s) is down', function(){
      var el = elem();
      var k = dispatcher(el);
      var mods = ['ctrl', 'shift', 'command', 'alt'];
      var unpress;
      for (var i = 0; i < mods.length; ++i) {
        unpress = press(el, mods[i]);
        assert(true == k.modifiers);
        unpress();
        assert(null == k.modifiers);
      }
    })

    it('should be `falsey` if no modifiers are down', function(){
      var el = elem();
      var k = dispatcher(el);
      assert(null == k.modifiers);
      var shift = press(el, 'shift');
      var ctrl = press(el, 'ctrl');
      assert(true == k.modifiers);
      shift();
      assert(true == k.modifiers);
      ctrl();
      assert(null == k.modifiers);
    })

    it('should be `falsey` after regaining focus', function() {
      var el = elem();
      var k = dispatcher(el);
      assert(null == k.modifiers);
      assert(null == k.ctrl);
      press(el, 'ctrl');
      assert(true == k.modifiers);
      assert(true == k.ctrl);
      el.dispatchEvent(new FocusEvent('focus'));
      assert(null == k.modifiers);
      assert(null == k.ctrl);
    })
  })


  // press `el` with `code`

  function press(el, code){
    if (!el || !code) throw new Error('(el, code) must be given to press()');
    if ('string' == typeof code) code = keycode(code);
    var e = document.createEvent('Event');
    e.initEvent('keydown', true, true);
    e.keyCode = e.which = code;
    el.dispatchEvent(e);
    return function(){
      var e = document.createEvent('Event');
      e.initEvent('keyup', true, true);
      e.keyCode = e.which = code;
      el.dispatchEvent(e);
    };
  }

  // create element with `type`

  function elem(type){
    return document.createElement(type || 'div');
  }
})


lib/proto.js

/**
 * Module Dependencies.
 */

var sequence = require('k-sequence');
var keycode = require('keycode');
var event = require('event');
var os = require('os');

/**
 * modifiers.
 */

var modifiers = {
  224: 'command',
  91: 'command',
  93: 'command',
  16: 'shift',
  17: 'ctrl',
  18: 'alt'
};

/**
 * Super key.
 * (must use subscript vs. dot notation to avoid issues with older browsers)
 */

exports['super'] = 'mac' == os
  ? 'command'
  : 'ctrl';

/**
 * Handle the given `KeyboardEvent` or bind
 * a new `keys` handler.
 *
 * @param {String|KeyboardEvent} e
 * @param {Function} fn
 * @api private
 */

exports.handle = function(e, fn){
  var ignore = this.ignore;
  var event = e.type;
  var code = e.which;

  // bind
  if (fn) return this.bind(e, fn);

  // modifiers
  var mod = modifiers[code];
  if ('keydown' == event && mod) {
    this['super'] = exports['super'] == mod;
    this[mod] = true;
    this.modifiers = true;
    this.active++;
    return;
  }

  // ignore
  if (ignore && ignore(e)) return;

  // listeners
  var all = this.listeners;

  // match
  for (var i = 0; i < all.length; ++i) {
    var invoke = true;
    var obj = all[i];
    var seq = obj.seq;
    var mods = obj.mods;
    var fn = seq || obj.fn;

    if (!seq && code != obj.code) continue;
    if (event != obj.event) continue;
    if (this.active != obj.mods.length) continue;

    for (var j = 0; j < mods.length; ++j) {
      if (!this[mods[j]]) {
        invoke = null;
        break;
      }
    }

    invoke && fn(e);
  }
};

/**
 * Destroy this `k` dispatcher instance.
 *
 * @api public
 */

exports.destroy = function(){
  event.unbind(this.el, 'keydown', this._handle);
  event.unbind(this.el, 'keyup', this._handle);
  event.unbind(this.el, 'keyup', this._clear);
  event.unbind(this.el, 'focus', this._clear);
  this.listeners = [];
};

/**
 * Unbind the given `keys` with optional `fn`.
 *
 * Example:
 *
 *      k.unbind('enter, tab', myListener); // unbind `myListener` from `enter, tab` keys
 *      k.unbind('enter, tab'); // unbind all `enter, tab` listeners
 *      k.unbind(); // unbind all listeners
 *
 * @param {String} keys
 * @param {Function} fn
 * @return {k}
 * @api public
 */

exports.unbind = function(keys, fn){
  var fns = this.listeners
  var len = fns.length;
  var all;

  // unbind all
  if (0 == arguments.length) {
    this.listeners = [];
    return this;
  }

  // parse
  all = parseKeys(keys);

  // unbind
  for (var i = 0; i < all.length; ++i) {
    for (var j = 0, obj; j < len; ++j) {
      obj = fns[j];
      if (!obj) continue;
      if (fn && obj.fn != fn) continue;
      if (obj.key != all[i].key) continue;
      if (!matches(obj, all[i])) continue;
      fns.splice(j--, 1);
    }
  }

  return this;
};

/**
 * Bind the given `keys` to `fn` with optional `event`
 *
 * Example:
 *
 *      k.bind('shift + tab, ctrl + a', function(e){});
 *
 * @param {String} event
 * @param {String} keys
 * @param {Function} fn
 * @return {k}
 * @api public
 */

exports.bind = function(event, keys, fn){
  var fns = this.listeners
  var len;
  var all;

  if (2 == arguments.length) {
    fn = keys;
    keys = event;
    event = 'keydown';
  }

  all = parseKeys(keys);
  len = all.length;

  for (var i = 0; i < len; ++i) {
    var obj = all[i];
    obj.seq = obj.seq && sequence(obj.key, fn);
    obj.event = event;
    obj.fn = fn;
    fns.push(obj);
  }

  return this;
};

/**
 * Bind keyup with `keys` and `fn`.
 *
 * @param {String} keys
 * @param {Function} fn
 * @return {k}
 * @api public
 */

exports.up = function(keys, fn){
  return this.bind('keyup', keys, fn);
};

/**
 * Bind keydown with `keys` and `fn`.
 *
 * @param {String} keys
 * @param {Function} fn
 * @return {k}
 * @api public
 */

exports.down = function(keys, fn){
  return this.bind('keydown', keys, fn);
};

/**
 * Clear all modifiers on `keyup`.
 *
 * @api private
 */

exports.clear = function(e){
  var code = e.keyCode || e.which;
  if (!(code in modifiers)) return;
  this.active--;
  this[modifiers[code]] = null;
  this.modifiers = this.command
    || this.shift
    || this.ctrl
    || this.alt;
};

/**
 * Clear all modifiers on `focus`.
 *
 * @api private
 */

exports.reset = function(e){
  this.active = 0;
  this.modifiers =
  this.command =
  this.shift =
  this.ctrl =
  this.alt = null;
};

/**
 * Ignore all input elements by default.
 *
 * @param {Event} e
 * @return {Boolean}
 * @api private
 */

exports.ignore = function(e){
  var el = e.target || e.srcElement;
  var name = el.tagName.toLowerCase();
  return 'textarea' == name
    || 'select' == name
    || 'input' == name;
};

/**
 * Parse the given `keys`.
 *
 * @param {String} keys
 * @return {Array}
 * @api private
 */

function parseKeys(keys){
  keys = keys.replace('super', exports['super']);

  var all = ',' != keys
    ? keys.split(/ *, */)
    : [','];

  var ret = [];
  for (var i = 0; i < all.length; ++i) {
    if ('' == all[i]) continue;
    var mods = all[i].split(/ *\+ */);
    var key = mods.pop() || ',';

    ret.push({
      seq: !! (~key.indexOf(' ') || ~key.indexOf('*')),
      code: keycode(key),
      mods: mods,
      key: key
    });
  }

  return ret;
}

/**
 * Check if the given `a` matches `b`.
 *
 * @param {Object} a
 * @param {Object} b
 * @return {Boolean}
 * @api private
 */

function matches(a, b){
  return 0 == b.mods.length || eql(a, b);
}

/**
 * Shallow eql util.
 *
 * TODO: move to yields/eql
 *
 * @param {Array} a
 * @param {Array} b
 * @return {Boolean}
 * @api private
 */

function eql(a, b){
  a = a.mods.sort().toString();
  b = b.mods.sort().toString();
  return a == b;
}
